File: src/main/java/com/lessmarkup/interfaces/data/ChangeTracker.java
package com.lessmarkup.interfaces.data;

import com.lessmarkup.interfaces.cache.EntityChangeType;

public interface ChangeTracker {
    void invalidate();
    <T extends DataObject> void addChange(Class<T> type, long objectId, EntityChangeType changeType, DomainModel domainModel);
    <T extends DataObject> void addChange(Class<T> type, T dataObject, EntityChangeType changeType, DomainModel domainModel);
    void subscribe(ChangeListener listener);
    void unsubscribe(ChangeListener listener);
    void enqueueUpdates();
}

File: src/main/scala/com/lessmarkup/interfaces/data/ChangeTracker.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.data

import com.lessmarkup.interfaces.cache.EntityChangeType

trait ChangeTracker {
  def invalidate()

  def addChange[T <: DataObject](`type`: Class[T], objectId: Long, changeType: EntityChangeType, domainModel: DomainModel)

  def addChange[T <: DataObject](`type`: Class[T], dataObject: T, changeType: EntityChangeType, domainModel: DomainModel)

  def subscribe(listener: ChangeListener)

  def unsubscribe(listener: ChangeListener)

  def enqueueUpdates()
}
